File: day07/part2/main.dart
// -*- compile-command: "dart main.dart ../example1.txt ../example2.txt ../input.txt"; -*-

import 'dart:io';

main(List<String> args) {
  for (var arg in args) {
    process(arg);
  }

  print('Done.');
}

class Contains {
  int quant;
  String color;
  Contains({this.quant, this.color});

  toString() => 'Contains(quant: $quant, color: $color)';
}

process(String filename) {
  print('Processing $filename ...');

  var lineRE = RegExp(r'^(.*?) bags contain (.*)\.$');
  var containsRE = RegExp(r',?\s*(\d+) (.*?) bags?');

  var rules = Map<String,List<Contains>>();

  var lines = File(filename).readAsLinesSync();
  for (var line in lines) {
    var m = lineRE.firstMatch(line);
    // print('${m.group(0)}; ${m.group(1)}; ${m.group(2)}');
    if (m.group(2) == 'no other bags') {
      rules[m.group(1)] = [];  // empty list
      // print('No bags: rules[${m.group(1)}] = []');
      continue;
    }

    var m2 = containsRE.allMatches(m.group(2));
    m2.forEach((f) {
        var q = int.parse(f.group(1));
        // print('${f.group(0)}; ${f.group(1)}; ${f.group(2)}');
        var c = Contains(quant: q, color: f.group(2));
        // if (rules.containsKey(m.group(1))) {
        //   print('BEFORE: rules[${m.group(1)}] = ${rules[m.group(1)]}');
        // } else {
        //   print('BEFORE: rules[${m.group(1)}] = null');
        // }
        rules.update(m.group(1), (cs) {
            // print('CS BEFORE: cs=$cs');
            cs.add(c);
            // print('CS AFTER: cs=$cs');
            return cs;  // CRITICAL TO PROVIDE!!!
          }, ifAbsent: () => [c]);
        // print('AFTER: rules[${m.group(1)}] = ${rules[m.group(1)]}');
    });
  }

  var count = contains("shiny gold", rules);

  print('Solution: ${count-1}');
}

int contains(String key, Map<String,List<Contains>> rules) {
  var total = 1;

  if (!rules.containsKey(key) || rules[key] == null) return 1;

  rules[key].forEach((v) {
      var count = contains(v.color, rules);
      total += (count * v.quant);
  });

  return total;
}

File: day07/part2/main.go
// -*- compile-command: "go run main.go ../example1.txt ../example2.txt ../input.txt"; -*-

package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"regexp"
	"strconv"
	"strings"
)

var (
	verbose = flag.Bool("v", false, "Verbose log messages")

	lineRE     = regexp.MustCompile(`^(.*?) bags contain (.*)\.$`)
	containsRE = regexp.MustCompile(`,?\s*(\d+) (.*?) bags?`)
)

func main() {
	flag.Parse()

	for _, arg := range flag.Args() {
		process(arg)
	}

	log.Printf("Done.")
}

type Contains struct {
	quant int
	color string
}

func process(filename string) {
	rules := map[string][]*Contains{}

	lines := readLines(filename)
	// log.Printf("lines=%v", lines)
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		m := lineRE.FindStringSubmatch(line)
		if len(m) != 3 {
			log.Fatalf("bad parse: %v", line)
		}
		// log.Printf("len(m)=%v: %v\n%#v", len(m), line, m)
		if m[2] == "no other bags" {
			rules[m[1]] = nil
			continue
		}

		m2 := containsRE.FindAllStringSubmatch(m[2], -1)
		// log.Printf("len(m2)=%v: %v\n%#v", len(m2), m[2], m2)
		for _, f := range m2 {
			q, err := strconv.Atoi(f[1])
			if err != nil {
				log.Fatalf("strconv.Atoi(%q): %v", f[1], err)
			}
			rules[m[1]] = append(rules[m[1]], &Contains{quant: q, color: f[2]})
		}
	}

	count := contains("shiny gold", rules, nil)

	fmt.Printf("Solution: %v\n", count-1)
}

func contains(key string, rules map[string][]*Contains, seen map[string]bool) int {
	total := 1
	for _, v := range rules[key] {
		count := contains(v.color, rules, seen)
		total += (count * v.quant)
	}
	// log.Printf("contains(%q)=%v", key, total)
	return total
}

func readLines(filename string) []string {
	log.Printf("Processing %v ...", filename)
	buf, err := ioutil.ReadFile(filename)
	check("ReadFile: %v", err)
	lines := strings.Split(string(buf), "\n")
	return lines
}

func check(fmtStr string, args ...interface{}) {
	if err := args[len(args)-1]; err != nil {
		log.Fatalf(fmtStr, args...)
	}
}

func logf(fmt string, args ...interface{}) {
	if *verbose {
		log.Printf(fmt, args...)
	}
}
